### ISSUE TITLE
Bug: Half-Life MDL broken bone hierarchy

### ISSUE THREAD

**Issue by turol** [assimp/assimp#5002]
**Describe the bug**
Model `test/models/MDL/MDL (HL1)/man.mdl` has hierarchy like this:

```
<MDL_root>
├╴<MDL_bones>
│ ├╴Root
│ ├╴Pelvis
│ ├╴RightLeg
│ ├╴LeftLeg
│ ├╴Spine
│ ├╴RightArm
│ ├╴LeftArm
│ └╴Neck
```

But the bone nodes have "MDL_root" as their parent instead of the correct "MDL_bones". In addition the hierarchy doesn't appear to match what's actually in the file. It looks like this:

```
0 "Root" parent -1
1 "Pelvis" parent 0
2 "RightLeg" parent 1
3 "LeftLeg" parent 1
4 "Spine" parent 1
5 "RightArm" parent 4
6 "LeftArm" parent 4
7 "Neck" parent 4
```

The hierarchy looks to have been flattened which would lead to wrong rendering of the model.

The code is here:
https://github.com/assimp/assimp/blob/f3767a4eb21f2f272102d0fe4020d4e4ca3154bc/code/AssetLib/MDL/HalfLife/HL1MDLLoader.cpp#L487
It incorrectly assigns the scene root `scene_->mRootNode` as the "Root" bone parent instead of "MDL_bones" node `bones_node`.
The child bones appear to get the correct parent but they are still in the root bone children array instead of being moved to their actual parent's children.

**To Reproduce**
https://github.com/assimp/assimp/pull/5001 fails unit tests because I added validation for the hierarchy. Grab that branch and run unit tests.

**Expected behavior**
Hierarchy is properly formed and the unit tests pass with the new validation.

@malortie I think you wrote this code. Can you fix it?

**Comment by malortie** [comment]
@turol 

> The hierarchy looks to have been flattened

By flattened, do you mean the following:

```
<MDL_bones>
├╴Root
├╴Pelvis
├╴RightLeg
├╴LeftLeg
├╴Spine
├╴RightArm
├╴LeftArm
└╴Neck
```

instead of:

```
<MDL_bones>
├╴Root
   ├╴Pelvis
      ├╴RightLeg
      ├╴LeftLeg
      ├╴Spine
         ├╴RightArm
         ├╴LeftArm
         └╴Neck
```

Yes, I think I originally made it that way so that each node (bone) could be accessed using indices in the same way it's done in HL1 code. But it's best to have it fixed so it works in conformance as the other importers.

I checked the code and I don't think there is a lot to change for now :

1. In `read_bones()`, `temp_bones_` is filled with all bones info and used in the rest of the code, so the parent/child hierarchy code that needs adjustment remains pretty much isolated inside this function.
2. The test [emptyBonesNames()](https://github.com/assimp/assimp/blob/0ed957f70a0bedc87431c41ac8381c8887b3531c/test/unit/ImportExport/MDL/utMDLImporter_HL1_Nodes.cpp#L80) in `utMDLImporter_HL1_Nodes.cpp` will have to be adjusted since the function [expect_named_children()](https://github.com/assimp/assimp/blob/0ed957f70a0bedc87431c41ac8381c8887b3531c/test/unit/ImportExport/MDL/utMDLImporter_HL1_Nodes.cpp#L97) assumes the node's children are flattened.

In addition, a new test should be added to ensure the new bone hierarchy is as expected.

Once done, I will update the [HL1 importer documentation](https://github.com/malortie/assimp/wiki/Assimp-Half-Life-1-MDL-importer).

I will try to look into it this weekend if I have time.

@kimkulling 
Because these changes will affect how parents/child nodes are stored, this will likely be a breaking change.

**Comment by turol** [comment]
Yes, that is the fixed hierarchy I had in mind.

**Comment by kimkulling** [comment]
Ok, so we need to increase the minor version number.

**Comment by malortie** [comment]
I created a new [branch](https://github.com/malortie/assimp/tree/validate_parents) with changes that fix the issues. The branch was created from #5001 and includes turol's changes in `ValidateDataStructure.cpp`. I can also create a separate PR with only HL1 MDL importer changes if preferable.

### Changes made to HL1 MDL Importer code

Hierarchy is now constructed to follow a tree-like structure (parent/child) instead of flattened. While reviewing the changes, I found that StudioMDL (the tool used to compile HL1 models) is able to compile SMD files that have multiple roots. The importer code now checks for root bones (parent: -1) and builds the hierarchy for each of them. They are now parented to the main 'bones' node `<MDL_bones>` instead of `scene->mRootNode`, as [previously explained](https://github.com/assimp/assimp/issues/5002#issue-1617193547).

I made a new model `multiple_roots.mdl` to use for testing the new hierarchy and to ensure correct loading of MDL models with multiple root bones. The SMD source files are available [here](https://github.com/malortie/assimp-hl1-mdl-importer-tests-sources/tree/master/models/multiple_roots).

### Changes made to unit tests

1 - A new test was added - `checkBoneHierarchy`. Uses `multiple_roots.mdl`.

2 - The test `emptyBonesNames` was adjusted to work with the new hierarchy.

3 - Various tests code were reworked. `expect_named_children()` was removed and tests that were using it now use `get_node_children_names()`. The new function is more flexible because it allows collecting node names that are under a specific hierarchy level. This is useful in tests checking only bodypart names but not their submodels. In addition, `expect_named_children()` was containing assertions checks. This is now decoupled with the new structure.

### Helpers

Various utilities were made/used to help test further. A separate [branch](https://github.com/malortie/assimp/tree/validate_parents_utils) has been made and contains additional utilities. The branch includes all changes from `validate_parents`, including changes for HL1 MDL importer.

> Note: CMake targets were added for each utility.

#### assview

Used to help visualize 3D MDL files. Only tested in Win32/x86.

- Adapted from [assview.c](https://github.com/ccxvii/asstools/blob/master/assview.c).
- [FreeGlut 3.0.0](https://www.transmissionzero.co.uk/software/freeglut-devel/) was used and must be installed manually to `samples/freeglut`.
- Changes were made to display line segments to be able to view bones.
- It does not handle loading embedded textures, so they won't appear in the viewer. Given no change was made to texture processing, this is likely not needed.

#### memory_leak_check

Used to check for memory leaks after importing a model and disposing the scene.

### Models tests

Each model below was loaded in assview. HL1 game specific models were additionally loaded in [HLAM](https://github.com/SamVanheer/HalfLifeAssetManager) to check correctness of their bone structure.

Half-Life 1 game models:

```
valve/models/
  player/
    barney.mdl
    gina.mdl
    gman.mdl
    gordon.mdl
    helmet.mdl
    hgrunt.mdl
    recon.mdl
    robo.mdl
    scientist.mdl
    zombie.mdl

  barney.mdl
  bullsquid.mdl
  hgrunt.mdl
  islave.mdl
  player.mdl
  scientist.mdl
  tentacle2.mdl
  v_9mmar.mdl
  w_9mmar.mdl
  zombie.mdl
```

Assimp tests models:

```
test/models/MDL/MDL (HL1)/
  duplicate_sequence_groups/duplicate_sequence_groups.mdl
  unnamed_sequence_groups/unnamed_sequence_groups.mdl

  alpha_test.mdl
  blend_additive.mdl
  chrome_sphere.mdl
  duplicate_bodyparts.mdl
  duplicate_sequences.mdl
  duplicate_submodels.mdl
  man.mdl
  multiple_roots.mdl
  sequence_transitions.mdl
  unnamed_bodyparts.mdl
  unnamed_bones.mdl
  unnamed_sequences.mdl
```

### Memory leaks

`memory_leak_check` was used to ensure no memory leaks were introduced with the new changes.

Memory leaks as such were found:

```
Detected memory leaks!
Dumping objects ->
{170} normal block at 0x007CE7E8, 8 bytes long.
 Data: <xF z    > 78 46 12 7A 00 00 00 00 
{169} normal block at 0x007CE238, 8 bytes long.
 Data: < E z    > 8C 45 12 7A 00 00 00 00 
{168} normal block at 0x007CE468, 8 bytes long.
 Data: < E z    > A8 45 12 7A 00 00 00 00 
{167} normal block at 0x007CE7B0, 8 bytes long.
 Data: < E z    > 20 45 12 7A 00 00 00 00 
{166} normal block at 0x007CE900, 12 bytes long.
 Data: <  |   |     > 00 E9 7C 00 00 E9 7C 00 CD CD CD CD 
{165} normal block at 0x007CE6D0, 8 bytes long.
 Data: <HE z    > 48 45 12 7A 00 00 00 00 
{164} normal block at 0x007C9C38, 28 bytes long.
 Data: <8 | 8 | 8 |     > 38 9C 7C 00 38 9C 7C 00 38 9C 7C 00 01 01 CD CD 
{163} normal block at 0x007CE388, 8 bytes long.
 Data: <<E z    > 3C 45 12 7A 00 00 00 00 
Object dump complete.
'memory_leak_checkd.exe' (Win32): Loaded 'C:\Windows\SysWOW64\kernel.appcore.dll'. 
The thread 0xdcfc has exited with code 0 (0x0).
The thread 0xa6e4 has exited with code 0 (0x0).
The program '[27868] memory_leak_checkd.exe' has exited with code 0 (0x0).
```

These occur even when `ASSIMP_BUILD_ALL_EXPORTERS_BY_DEFAULT` and `ASSIMP_BUILD_ALL_IMPORTERS_BY_DEFAULT` are disabled, and even when no scene is loaded. These could probably be related to #2082 . During testing, both were disabled and only `ASSIMP_BUILD_MDL_IMPORTER` was enabled.

I also loaded each aforementioned model and checked to see if additional blocks would show up. So far all block numbers appear to remain consistent and no additional block occur. The only notable things changing occasionally are the numbers in the span [163, ... , 170], sometimes starting at a higher number i.e. [164, ... , 171], however the span remains the same, so probably due to 2082 and not the HL1 MDL importer. This behavior also occurs in x64.

### Unit tests execution

Unit tests were run on Debug, MinSizeRel, Release, RelWithDebInfo.

- Win32: All tests pass, but `aiVector2DotProductTest` and `aiVector3DotProductTest` occasionally fail in Release and RelWithDebInfo.
- x64: All tests pass.

### Compiling

Compilation was tested in Win32 and x64, each on Debug, MinSizeRel, Release, RelWithDebInfo.

Everything builds correctly, as for the HL1 MDL importer changes.

> Note: `assview` is set to work with the Freeglut x86 libraries and will only compile on Win32.
>
> Note: Building only with `ASSIMP_BUILD_MDL_IMPORTER` enabled and all the other importers/exporters disabled will fail because [MDLImporter](https://github.com/assimp/assimp/blob/master/code/AssetLib/MDL/MDLLoader.cpp#L517) uses `LookupNormalIndex`, which is defined in [MD2Loader.cpp](https://github.com/assimp/assimp/blob/master/code/AssetLib/MD2/MD2Loader.cpp#L83).
>
> This can be solved by moving the function to a separate file shared between MD2 and MDL.

---

@turol Please, can you check the changes and confirm if it works ? Optionally, you may use the helpers in `validate_parents_utils` branch.

**Comment by turol** [comment]
@malortie Nice work. Are you going to create a PR or should I? Some notes:

My new validation change should not be in this PR. The goal is that every commit in history should build and pass unit tests. This makes it easier to bisect.

If you created new test models please add them and matching unit tests.

About assview
First of all, phrasing. We're already getting comments about the name of the project.
Second, we should have a real portable modern viewer that would replace the current DX9 one. It should use portable libraries for rendering and GUI. Probably imgui for GUI, modern OpenGL for rendering and SDL/GLFW for platform specifics. This should also be a separate PR.

Your leak checker appears to only work on win32/MSVC but the CMake files don't restrict it to that platform. It's also unnecessary since it's possible to use Asan/Valgrind with assimp command or unit tests and in fact the CI does just that. Yes, there are a lot of memory leaks. Separate PRs for fixing them are welcome.

**Comment by malortie** [comment]
I created the PR for HL1 MDL importer fixes. See https://github.com/assimp/assimp/pull/5007.

In regards to the helpers I made, those were made only for the purpose of helping debug and ensure the changes were correct so I will not make PRs for them.

I will update the HL1 MDL importer documentation once the PR is merged.

**Comment by malortie** [comment]
The HL1 MDL importer bones node [documentation](https://github.com/malortie/assimp/wiki/Assimp-Half-Life-1-MDL-importer#bones-node) has been updated.
